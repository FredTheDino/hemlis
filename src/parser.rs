#[cfg(test)]
mod tests {
    use insta::assert_snapshot;

    macro_rules! gen_parser {
        ($a:ident, $p:ident) => {
            pub fn $a<'t>(src: &'t str) -> String {
                pub fn inner<'t>(src: &'t str) -> Result<String, String> {
                    use crate::{ast::Ast, grammer, lexer};
                    use std::io::BufWriter;

                    let parser = grammer::$p::new();
                    let out = parser
                        .parse(
                            lexer::lex(&src)
                                .into_iter()
                                .map(|(token, span)| Ok((span.start, token?, span.end))),
                        )
                        .map_err(|x| format!("{:?}", x))?;

                    let mut buf = BufWriter::new(Vec::new());
                    out.show(0, &mut buf).map_err(|x| format!("{:?}", x))?;
                    let inner = buf.into_inner().map_err(|x| format!("{:?}", x))?;
                    String::from_utf8(inner).map_err(|x| format!("{:?}", x))
                }
                match inner(src) {
                    Ok(s) | Err(s) => s,
                }
            }
        };
    }

    gen_parser!(p_header, HeaderParser);
    gen_parser!(p_import, ImportParser);

    #[test]
    fn empty_string() {
        assert_snapshot!(p_header(""));
    }

    #[test]
    fn normal_definition() {
        assert_snapshot!(p_header("module A (a, b, c) where"));
    }

    #[test]
    fn everything_header() {
        assert_snapshot!(p_header(r#"
module A.B.C (a, class B, C, D(..), E(F, G), 
 (+), type (+), module H) where

import A as A
import A.B.C as A.C
import A.B.C (a, class B, 
 C, D(..), E(F, G), (+), type (+))
import A.B.C hiding (foo)
        "#));
    }

    #[test]
    fn simple_import() {
        assert_snapshot!(p_import("import A as A"))
    }

    #[test]
    fn import_with_newlines() {
        assert_snapshot!(p_import("import A (foo\n , bar\n , baz)"))
    }
}
