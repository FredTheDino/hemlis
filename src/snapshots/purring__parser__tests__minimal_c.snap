---
source: src/parser.rs
expression: "p_module(r\"\nmodule AccessRight where\n\nf x =\n  let\n    a b =\n      case b of\n        ABC { a } -> 1\n        _ -> 1\n  in 1\n            \")"
---
32 of 32
===
Module
 Header
  QProperName
   NULL
   ProperName
    8..19
  Decl::Def
   Name
    27..28
    Binder::Var
     Name
      29..30
   GuardedExpr::Unconditional
    Expr::Let
     35..38
      LetBinding::Name
       Name
        43..44
        Binder::Var
         Name
          45..46
       GuardedExpr::Unconditional
        Expr::Case
         55..59
          Expr::Ident
           QName
            NULL
            Name
             60..61
          CaseBranch
            Binder::App
             Binder::Constructor
              QProperName
               NULL
               ProperName
                73..76
             Binder::Record
               RecordLabelBinder::Pun
                Name
                 79..80
           GuardedExpr::Unconditional
            Expr::Number
             Number
              86..87
          CaseBranch
            Binder::Wildcard
             96..97
           GuardedExpr::Unconditional
            Expr::Number
             Number
              101..102
     Expr::Number
      Number
       108..109
===
